rust/worker/src/blockstore/arrow_blockfile/block/delta.rs

```diff
@@ -0,0 +1,382 @@
+
+        let prefix_data_size = self.get_prefix_size() + additional_prefix_size;
+        let key_data_size = self.get_key_size() + additional_key_size;
+        let value_data_size = self.get_value_size() + additional_value_size;
+
+        let prefix_offset_size = bit_util::round_upto_multiple_of_64((self.new_data.len() + 1) * 4);
+        let key_offset_size = self.offset_size_for_key_type(self.new_data.len(), key.into());
+        let value_offset_size = self.offset_size_for_value_type(self.new_data.len(), value.into());
+
+        let prefix_total_bytes =
+            bit_util::round_upto_multiple_of_64(prefix_data_size) + prefix_offset_size;
+        let key_total_bytes = bit_util::round_upto_multiple_of_64(key_data_size) + key_offset_size;
+        let value_total_bytes =
+            bit_util::round_upto_multiple_of_64(value_data_size) + value_offset_size;
+        let total_future_size = prefix_total_bytes + key_total_bytes + value_total_bytes;
+
+        total_future_size <= MAX_BLOCK_SIZE
+    }
+
+    fn offset_size_for_value_type(&self, item_count: usize, value_type: ValueType) -> usize {
+        match value_type {
+            ValueType::Int32Array | ValueType::String => {
+                bit_util::round_upto_multiple_of_64((item_count + 1) * 4)
+            }
+            _ => unimplemented!("Value type not implemented"),
+        }
+    }
+
+    fn offset_size_for_key_type(&self, item_count: usize, key_type: KeyType) -> usize {
+        match key_type {
+            KeyType::String => bit_util::round_upto_multiple_of_64((item_count + 1) * 4),
+            KeyType::Float => 0,
+            _ => unimplemented!("Key type not implemented"),
+        }
+    }
+
+    fn split(
+        &mut self,
+        key_type: KeyType,
+        value_type: ValueType,
+    ) -> (BlockfileKey, BTreeMap<BlockfileKey, Value>) {
+        let half_size = MAX_BLOCK_SIZE / 2;
+        let mut running_prefix_size = 0;
+        let mut running_key_size = 0;
+        let mut running_value_size = 0;
+        let mut running_count = 0;
+        let mut split_key = None;
+        // The split key will be the last key that pushes the block over the half size. Not the first key that pushes it over
+        for (key, value) in self.new_data.iter() {
+            running_prefix_size += key.get_prefix_size();
+            running_key_size += key.key.get_size();
+            running_value_size += value.get_size();
+            running_count += 1;
+            let current_size = self.get_block_size(
+                running_count,
+                running_prefix_size,
+                running_key_size,
+                running_value_size,
+                key_type,
+                value_type,
+            );
+            if half_size < current_size {
+                break;
+            }
+            split_key = Some(key.clone());
+        }
+
+        match &split_key {
+            None => panic!("No split point found"),
```

Comment:
Can we add a TODO here to return a NotBigEnoughToSplitError or something

Reply:
I was torn on this - this should never really happen. Since above code should always only split if you can be split. That is why I left it as a panic, it represents an invariant in the code, "one should always check that it can be split before calling this" I can make it an error or document it better. Wdyt?

Reply:
Ah I see. In that case I'm fine with this panicking at least for now. We'll probably eventually want to error it, monitor the frequency of the error, and page if it ever goes above 0. But for now, panic :)

Reply:
Added some documentation and a note